static/js/blueslip_stacktrace.ts

```diff
@@ -0,0 +1,122 @@
+    } catch {
+        full_fn_name = data.functionName;
+    }
+    const fn_name = full_fn_name.split('.');
+
+    return {
+        full_path: full_path,
+        show_path: show_path,
+        line_number: data.lineNumber,
+        function_name: fn_name.pop(),
+        function_scope: fn_name.join('.'),
+    };
+}
+
+// The stacktrace-gps library relies on the mozilla/source-map WASM library,
+// we assume the developer's browser is recent enough to support it.
+const stack_trace_gps = new StackTraceGPS();
+
+export async function display_stacktrace(error: string, stack: string): Promise<void> {
+    const ex = new Error();
+    ex.stack = stack;
+
+    const stackframes: CleanStackFrame[] = await Promise.all(
+        ErrorStackParser.parse(ex).map(async (stackframe: StackFrame) => {
+            const mapped_stackframe = await stack_trace_gps.getMappedLocation(stackframe);
+            return clean_stackframe_data(mapped_stackframe);
+        })
+    );
+
+    const rendered = render_blueslip_stacktrace({ stackframes });
+
+    $('.stacktrace .stacktrace-content').html(rendered);
+    $('.stacktrace .message span').text(error);
+    $('.stacktrace').addClass('show');
+}
+
+function expand_code_context(target: JQuery): void {
+    if (target.css('maxHeight') !== '0px') {
+        target.css('maxHeight', '0px');
+    } else {
+        target.css('maxHeight', target[0].scrollHeight + 'px');
+    }
+    target.toggleClass('show');
+}
+
+export function display_code_context(
+    full_path: string,
+    line_number: number,
+    context_wrapper: JQuery
+): void {
+    if (context_wrapper.children().first().is(':empty')) {
+        $.get('/webpack/app.js.map', async (map: string) => {
+            const sourceContent = await SourceMapConsumer.with(
+                map, null, async (consumer: IndexedSourceMapConsumer) => {
+                    // We don't know the exact file path due to the node_modules
+                    // hash so we have to find it.
+                    for (const source of consumer.sources) {
+                        if (source.match(full_path + '$')) {
+                            return consumer.sourceContentFor(source);
+                        }
+                    }
+                    throw Error('Cannot find module');
+                }
+            );
+
+            const lines = sourceContent.split('\n');
+            let lo_line_num = Math.max(line_number - 5, 0);
+            const hi_line_num = Math.min(line_number + 4, lines.length);
+            const rendered_context = lines.slice(lo_line_num, hi_line_num)
+                .map((line: string) => {
+                    line = line.replace(/\t/g, '    ');
+                    lo_line_num += 1;
+                    return '<div class="' + (lo_line_num === line_number ? 'focus-line' : '') + '">'
+                        + '<span class="line-number">' + lo_line_num + '</span>&nbsp;'
+                        + line + '</div>';
```

Comment:
As discussed before, do not try to generate HTML with string concatenation. This doesn’t escape anything, so it’s wrong. It may or may not be a security vulnerability depending on whether an attacker can find a way to control the input, but it’s wrong either way.